crates/ntx-builder/src/builder/mod.rs

```diff
@@ -1,95 +1,305 @@
+            notes.iter().cloned().map(Note::from).map(InputNote::unauthenticated).collect(),
+        )
+        .context("failed to create InputNotes object")?;
+
+        for n in input_notes.iter() {
+            data_store.insert_note_script_mast(n.note().script());
+        }
+
+        let checker = NoteConsumptionChecker::new(tx_executor);
+        match checker
+            .check_notes_consumability(
+                account.id(),
+                block_ref,
+                input_notes.clone(),
+                TransactionArgs::default(),
+                Arc::new(DefaultSourceManager::default()),
+            )
+            .await
+        {
+            Ok(NoteAccountExecution::Success) => input_notes,
+            Ok(NoteAccountExecution::Failure { successful_notes, .. }) => {
+                // Take only successful notes, put failing notes back to the queue
+                let successul_notes =
+                    input_notes.iter().filter(|n| successful_notes.contains(&n.id())).cloned();
+
+                let unsuccessful_notes: Vec<NetworkNote> = input_notes
+                    .iter()
+                    .filter(|n| !successful_notes.contains(&n.id()))
+                    .map(InputNote::note)
+                    .cloned()
+                    .map(|n| NetworkNote::try_from(n).expect("checked above"))
+                    .collect();
+
+                api_state
+                    .lock()
+                    .await
+                    .queue_unconsumed_notes(unsuccessful_notes.into_iter().rev());
+
+                InputNotes::new(successul_notes.collect()).unwrap()
+            },
+            Err(err) => {
+                warn!(target: COMPONENT, error=%err, "Consumption check failed");
+                api_state.lock().await.queue_unconsumed_notes(notes);
+                data_store.evict_account(account.id());
+                return Ok(None);
+            },
+        }
+    };
+
+    if input_notes.iter().count() == 0 {
+        warn!(target: COMPONENT, "No notes left to be consumed after consumption check");
+
+        return Ok(None);
+    }
+
+    match tx_executor
+        .execute_transaction(
+            account.id(),
+            block_ref,
+            input_notes,
+            TransactionArgs::default(),
+            Arc::new(DefaultSourceManager::default()),
+        )
+        .await
+    {
+        Ok(tx) => Ok(Some(tx)),
+        Err(err) => {
+            warn!(target: COMPONENT, error=%err, "Network transaction execution failed");
+            api_state.lock().await.queue_unconsumed_notes(notes);
+            data_store.evict_account(account.id());
+
+            Ok(None)
+        },
+    }
+}
+
+/// Proves and submit an executed transaction
+async fn prove_and_submit(
+    api_state: &SharedPendingNotes,
+    block_producer_client: &BlockProducerClient,
+    executed_tx: &ExecutedTransaction,
+    remote_tx_prover: Option<&RemoteTransactionProver>,
+) -> anyhow::Result<()> {
```

Comment:
I would maybe create an enum - something like `TransactionProver` which could look like so:
```rust
enum TransactionProver {
    Local(LocalTransactionProver),
    Remote(RemoteTransactionProver),
}
```
And then this function would be a method on this struct.